### PR TITLE
expression: rewrite builtin function: DATEDIFF

### PR DESCRIPTION
**Pull request by spongedu**
for #4080

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -260,42 +260,34 @@ func (c *dateDiffFunctionClass) getFunction(args []Expression, ctx context.Conte
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinDateDiffSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpTime, tpTime)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
```
Comment (zz-jason) [inline expression/builtin_time.go]: set `Flen` and `Decimal` of `bf.tp`

File: expression/builtin_time.go [hunk 2]
```diff
@@ -260,42 +260,34 @@ func (c *dateDiffFunctionClass) getFunction(args []Expression, ctx context.Conte
...
-		return
-	}
-	sc := b.ctx.GetSessionVars().StmtCtx
-	t1, err := convertDatumToTime(sc, args[0])
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinDateDiffSig) evalInt(row []types.Datum) (int64, bool, error) {
+	ctx := b.ctx.GetSessionVars().StmtCtx
+	t1, isNull, err := b.args[0].EvalTime(row, ctx)
+	if isNull || err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]: in mysql:
```sql
MySQL > select datediff("2014-12-09", "aaaa");
Field   1:  `datediff("2014-12-09", "aaaa")`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       LONGLONG
Collation:  binary (63)
Length:     7
Max_length: 0
Decimals:   0
Flags:      BINARY NUM


+--------------------------------+
| datediff("2014-12-09", "aaaa") |
+--------------------------------+
|                           NULL |
+--------------------------------+
1 row in set, 1 warning (0.00 sec)

MySQL > show warnings;
Field   1:  `Level`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8_general_ci (33)
Length:     21
Max_length: 7
Decimals:   31
Flags:      NOT_NULL

Field   2:  `Code`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       LONG
Collation:  binary (63)
Length:     4
Max_length: 4
Decimals:   0
Flags:      NOT_NULL UNSIGNED BINARY NUM

Field   3:  `Message`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8_general_ci (33)
Length:     1536
Max_length: 32
Decimals:   31
Flags:      NOT_NULL


+---------+------+----------------------------------+
| Level   | Code | Message                          |
+---------+------+----------------------------------+
| Warning | 1292 | Incorrect datetime value: 'aaaa' |
+---------+------+----------------------------------+
1 row in set (0.00 sec)
```

File: expression/integration_test.go [hunk 3]
```diff
@@ -963,6 +963,9 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
 	result.Check(testkit.Rows("62966505600 63426672000 63426721412 63426721412"))
 	result = tk.MustQuery("select to_days(950501), to_days('2007-10-07'), to_days('2007-10-07 00:00:59'), to_days('0000-01-01')")
 	result.Check(testkit.Rows("728779 733321 733321 1"))
+	result = tk.MustQuery("select datediff('2007-12-31 23:59:59','2007-12-30'), datediff('2010-11-30 23:59:59', " +
```
Comment (zz-jason) [inline expression/integration_test.go]: add more tests like `datediff("2014-12-09", "aaaa");`

File: plan/typeinfer_test.go [hunk 4]
```diff
@@ -994,5 +994,12 @@ func (s *testPlanSuite) createTestCase4TimeFuncs() []typeInferTestCase {
 		{"microsecond(c_blob     )", mysql.TypeLonglong, charset.CharsetBin, mysql.BinaryFlag, 6, 0},
 		{"microsecond(c_set      )", mysql.TypeLonglong, charset.CharsetBin, mysql.BinaryFlag, 6, 0},
 		{"microsecond(c_enum     )", mysql.TypeLonglong, charset.CharsetBin, mysql.BinaryFlag, 6, 0},
+
+		{"datediff(c_char, c_datetime)", mysql.TypeLonglong, charset.CharsetBin, mysql.BinaryFlag, 20, 0},
```
Comment (zz-jason) [inline plan/typeinfer_test.go]: `Flen` is not correct

File: expression/integration_test.go (hunk 3 above)
Comment (winkyao) [inline expression/integration_test.go]: add case like `select datediff('2007-12-31 23:59:59', '23:59:59');` .

Comment (spongedu) [inline expression/integration_test.go]: done

Comment (spongedu) [inline expression/integration_test.go]: done

File: plan/typeinfer_test.go (hunk 4 above)
Comment (spongedu) [inline plan/typeinfer_test.go]: Indeed it's not consistent with mysql. But I think it's reasonable to set it to mysql.MaxIntWidth. I'll take a look how this was set in mysql......

**Comment by zz-jason** [comment]
@spongedu plz resolve conflicts

**Comment by XuHuaiyu** [comment]
PTAL @winkyao @zz-jason

**Comment by zz-jason** [comment]
@XuHuaiyu @winkyao PTAL

**Comment by zz-jason** [comment]
@spongedu please resolve conflicts

File: expression/integration_test.go [hunk 5]
```diff
@@ -1078,6 +1078,9 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
...
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: add test cases like
1. "0000-01-01", "0001-01-01"
2. "0001-00-00", "0001-00-00"

Comment (spongedu) [inline expression/integration_test.go]: Done. @XuHuaiyu  PTAL

**Comment by XuHuaiyu** [comment]
Please merge the master, I've split tpTime into tpDatetime and tpTimestamp.

**Comment by spongedu** [comment]
@XuHuaiyu Done

**Comment by XuHuaiyu** [comment]
please fix ci

**Comment by spongedu** [comment]
@XuHuaiyu  Done. PTAL

**Pull request merged**